Make AIP and directory level metadata searchable

replaceafill opened the pull request:
This PR extracts descriptive and administrative metadata attached to directories through the `metadata.csv` file, the UI metadata forms or the `bag-info.txt` file and makes it searchable through the `transferMetadata` option in the Archival storage tab (the Any option will work too).

So far searching file metadata has relied on us indexing their `dmdSec` sections from the METS. But this presents several (potential) problems:

* See enhacement 3. in https://github.com/archivematica/Issues/issues/404
* https://github.com/artefactual/archivematica/pull/1463#issuecomment-536711483
* https://github.com/archivematica/Issues/issues/619#issuecomment-487190700

This PR also represents a step forward on deprecating the `METS` entry in the `aipfiles` index (the `METS` entry of the `aips` index was already removed in https://github.com/artefactual/archivematica/pull/1463) by:

* Creating a simple data structure with the descriptive metadata of the file. We are still missing a structure to index PREMIS data like events.
* Adding AIP metadata information to each entry in the `aipfiles` index. This may increase the size of the index but would allow us to avoid making separate queries when searching for AIPs. See enhancement 9. in https://github.com/archivematica/Issues/issues/404

Connected to https://github.com/archivematica/Issues/issues/547
Connected to https://github.com/archivematica/Issues/issues/888

sevein commented on src/archivematicaCommon/lib/elasticSearchFunctions.py:
```diff
@@ -457,20 +460,25 @@ def index_aip_and_files(
 
 
-def _index_aip_files(client, uuid, mets, name, identifiers=None):
+def _index_aip_files(client, uuid, mets, name, identifiers=None, aip_metadata=None):
```
Since we have a docstring describing all the existing parameters, could you add the new one to the list?

sevein commented on src/archivematicaCommon/lib/elasticSearchFunctions.py:
```diff
@@ -824,13 +841,131 @@ def _remove_tool_output_from_mets(doc):
-            doc, "mets:amdSec/mets:sourceMD/mets:mdWrap/mets:xmlData/transfer_metadata"
-        )
-    ]
+def get_directories_with_metadata(container):
```
I see a pattern in your PR where internal functions are not using the underscore convention, are you doing that on purpose because you think they should be public?

replaceafill replied:
No. Forgot to make them "private". Will fix.

sevein commented on src/archivematicaCommon/lib/elasticSearchFunctions.py:
```diff
@@ -824,13 +841,131 @@ def _remove_tool_output_from_mets(doc):
+    """
+    result = []
+    for directory in ns.xml_findall_premis(container, './/mets:div[@TYPE="Directory"]'):
+        if "DMDID" in directory.attrib or "ADMID" in directory.attrib:
```
Not sure how expensive `directory.attrib` is, but if it was, should be avoided?
Using an extra var? Or `directory.attrib in ('DMDID', 'AMDID')`?

replaceafill replied:
`directory.attrib` is a dictionary in this case. I really wanted to do it through a single XPath query but couldn't make it work. For example you could do:

```
ns.xml_findall_premis(container, './/mets:div[@TYPE="Directory"][@DMDID]')
```

and get all directories with descriptive metadata sections, but I couldn't make an `OR` work there, like:

```
ns.xml_findall_premis(container, './/mets:div[@TYPE="Directory"][@DMDID or @AMDID ]')
```

replaceafill replied:
Thanks to @jraddaoui's help I made the `OR` work in https://github.com/artefactual/archivematica/pull/1514/commits/220c93c5e0a34943449c6ff4411ed718622c8fed :bowing_man:

sevein commented on src/archivematicaCommon/lib/elasticSearchFunctions.py:
```diff
@@ -824,13 +841,131 @@ def _remove_tool_output_from_mets(doc):
+            directory_metadata = get_directory_metadata(directory, doc)
+            if directory_metadata:
+                result.append(directory_metadata)
+    return result
```
I'm wondering if this would be a good case for generators. Say you're looking at 10k directories, would generators help noticeably? If it's not much of an effort (yields and yields), perhaps we should?

replaceafill replied:
Thanks for the idea. `_get_directories_with_metadata` can be converted to a generator :+1:

replaceafill replied:
`_get_directories_with_metadata` ended as a single call to lxml's xpath.

sevein reviewed (commented):
I've submitted a few comments. I haven't tested yet, will do that next.

replaceafill commented:
@sevein I think I addressed your observations. Let me know how testing goes.

sevein reviewed (approved):
Thanks, Douglas. I've tested this and the limitations described around searches now seem to be solved. It also appears that you've made an effort to not introduce backward-incompatible changes in the searches, i.e. users won't need to rebuild the search indices necessarily while they may still do if they wish.

It'd be interesting to identify what's missing in metsrw that it didn't make it appealing in this particular use case. If you have a clear idea, I'd suggest to file [a new issue](https://github.com/artefactual-labs/mets-reader-writer/issues/new) in the metsrw repo. We could try to complete metsrw for this particular use case at some point.

:clap: :clap: :clap:

replaceafill committed:
Make directory level metadata searchable

This extracts descriptive and administrative metadata attached to
directories through the metadata.csv file, the UI metadata forms or
the bag-info.txt file and makes it searchable through the
transferMetadata option in the Archival storage tab (the Any option
will work too).

This also represents a step forward on deprecating the METS entry in
the aipfiles index (the METS entry of the aips index was already
removed in archivematica/PR#1463) by:

* Creating a simple data structure with the descriptive metadata of
  the file. We are still missing a structure to index PREMIS data like
  events.

* Adding AIP metadata information to each entry in the aipfiles
  index. This may increase the size of the index but would allow us to
  avoid making separate queries when searching for AIPs. See
  enhancement 9. in archivematica/Issues#404

ross-spencer commented:
Hi @sevein @replaceafill I feel the issue should be https://github.com/archivematica/issues still, but the potential use case does sound good. We haven't a label in issues yet for `metsrw` I believe, but we can add one if it helps tease apart AM core from other libs.
